metals/src/main/scala/scala/meta/internal/metals/CompilationScheduler.scala

```diff
@@ -0,0 +1,108 @@
+package scala.meta.internal.metals
+import ch.epfl.scala.{bsp4j => b}
+import scala.collection.concurrent.TrieMap
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.io.AbsolutePath
+
+final class CompilationScheduler(
+    buildTargets: BuildTargets,
+    classes: BuildTargetClasses,
+    workspace: () => AbsolutePath,
+    buildServer: () => Option[BuildServerConnection]
+)(implicit ec: ExecutionContext) {
+  private val batch = new BatchedFunction[Request, b.CompileResult](compile)
```

Comment:
As discussed offline, there should be two separate compile queues

- regular compiles for `didSave`
- cascade compiles for miscellaneous use-cases (import build, find references, manual invocation)